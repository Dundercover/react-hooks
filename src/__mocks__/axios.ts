import { Payload } from '../services/api'

const data: Payload = {
  type: 'number',
  length: 2,
  data: [5],
  success: true,
}

export default {
  get: jest.fn(() => Promise.resolve({ data })),
}
