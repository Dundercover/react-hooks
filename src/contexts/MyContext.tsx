import React, {
  useCallback,
  useEffect,
  useMemo,
  useState,
  useContext,
} from 'react'
import * as api from '../services/api'
import useRandomNumber from '../hooks/useRandomNumber'

type State = { isLoading: boolean; value: number }
type Actions = {
  fetchDataAsync: () => void
  increment: () => void
  reset: () => void
  setValue: (value: number) => void
}

const MyContext = React.createContext({} as [State, Actions])

const MyContextProvider: React.FunctionComponent = ({ children }) => {
  const [isLoading, setLoading] = useState(true)
  const { value, increment, reset, setValue } = useRandomNumber()

  const controller = useMemo(() => new AbortController(), [])

  const fetchDataAsync = useCallback(async () => {
    const { signal } = controller
    setLoading(true)
    const { data, length } = await api.fetchRandomNumberAsync(signal)
    if (length > 0) {
      setValue(data[0])
    }
    setLoading(false)
  }, [])

  useEffect(() => {
    fetchDataAsync()
    return () => controller.abort()
  }, [])
  return (
    <MyContext.Provider
      value={[
        { isLoading, value },
        {
          fetchDataAsync,
          increment,
          reset,
          setValue,
        },
      ]}>
      {children}
    </MyContext.Provider>
  )
}

const MyContextConsumer = MyContext.Consumer

export { MyContext, MyContextProvider, MyContextConsumer }
