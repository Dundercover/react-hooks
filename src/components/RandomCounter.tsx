import React, { useContext, useCallback, useMemo } from 'react'
import { StateContext, ActionsContext } from '../contexts/MyContext'
import styled from 'styled-components'

const StyledButton = styled.button``

const RandomCounter: React.FunctionComponent = ({ children }) => {
  const context = useContext(StateContext)
  const actions = useContext(ActionsContext)

  const handleChange = useCallback(
    (event: React.ChangeEvent<HTMLInputElement>) =>
      actions.setValue(+event.target.value),
    [actions]
  )
  const value = useMemo(() => (context.isLoading ? '...' : context.value), [
    context.isLoading,
    context.value,
  ])
  return (
    <>
      <div>{`Value: ${value}`}</div>
      <StyledButton disabled={context.isLoading} onClick={actions.increment}>
        Increment
      </StyledButton>
      <StyledButton disabled={context.isLoading} onClick={actions.reset}>
        Clear
      </StyledButton>
      <StyledButton
        disabled={context.isLoading}
        onClick={actions.fetchDataAsync}>
        Fetch
      </StyledButton>
      <input
        aria-label="input-field"
        disabled={context.isLoading}
        value={value}
        onChange={handleChange}
      />
      {context.isLoading && <div>is loading</div>}
    </>
  )
}

export default RandomCounter
