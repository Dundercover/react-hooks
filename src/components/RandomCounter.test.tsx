import 'abortcontroller-polyfill/dist/polyfill-patch-fetch'
import 'jest-dom/extend-expect'
import React from 'react'
import {
  render,
  fireEvent,
  cleanup,
  waitForElement,
  waitForDomChange,
} from 'react-testing-library'
import RandomCounter from './RandomCounter'
import { MyContextProvider } from '../contexts/MyContext'
import axios from 'axios'

afterEach(cleanup)

test('generates a random number on init', async () => {
  const { getByText, getByLabelText, container } = render(
    <MyContextProvider>
      <RandomCounter />
    </MyContextProvider>
  )
  const input = getByLabelText('input-field') as HTMLInputElement
  expect(input.value).toBe('...')
  expect(input.disabled).toBe(true)
  expect(container.firstChild).toMatchSnapshot()

  await waitForDomChange()
  expect(input.value).toBe('5')
  expect(input.disabled).toBe(false)

  expect(axios.get).toHaveBeenCalledTimes(1)
})
