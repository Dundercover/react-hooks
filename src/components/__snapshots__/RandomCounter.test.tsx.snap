// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generates a random number on init 1`] = `
<div>
  ...
</div>
`;
